### PR TITLE
NAV - Testing Kalman Filter

### PR DESCRIPTION
**Pull request by misha7b**
TODO:
- Work out how to actually run test
- Test that Kalman Filter runs
- Test that Kalman Filter runs correctly by going through calculations on paper

### CODE REVIEW

File: lib/navigation/filtering/kalman_filter.cpp [hunk 1]
```diff
@@ -22,18 +22,39 @@ KalmanFilter::KalmanFilter(const StateVector &initial_state,
...
+  // x_k = x_k + K_k*(z_k - H*x_k)
+
   state_estimate_
-    = prop_state_estimate + kalman_gain * (measurement - measurement_matrix * prop_state_estimate);
+    = state_estimate_ + kalman_gain * (measurement - measurement_matrix * state_estimate_);
+
+  // P_k = (I - K_k*H)*P_k*(I - K_k*H)^T + K_k*R*K_k^T
+
+  auto difference = (Eigen::Matrix<core::Float, state_dimension, state_dimension>::Identity()
+                     - kalman_gain * measurement_matrix);
```
Comment (TomLonergan03) [inline lib/navigation/filtering/kalman_filter.cpp:54]: extraneous brackets

File: test/navigation/kalman_filter.cpp [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #include <iostream>
+#include <random>
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:2]: what is this used for?

File: test/navigation/kalman_filter.cpp [hunk 3]
```diff
@@ -10,21 +11,113 @@
 namespace hyped::test {
 
 TEST(KalmanFilter, construction)
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:13]: rename the test

File: test/navigation/kalman_filter.cpp [hunk 4]
```diff
@@ -10,21 +11,113 @@
...
+// Measurments taken every 1s
+// 20 measurments taken
+
+// Variacnce of measurments:
+//  Distance: 10
+//  Velocity: 5
+//  Acceleration: 3
+
+// Expected displacement: 4000m
+// Expected velocity: 400ms^-1
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:29]: move these comments above `TEST...`

File: test/navigation/kalman_filter.cpp [hunk 5]
```diff
@@ -10,21 +11,113 @@
...
+
+// Test simulating simple cart movement
+
+// Acceleration: 20ms^-2
+// Initial velocity: 0
+// Initial position: 0
+// Measurments taken every 1s
+// 20 measurments taken
+
+// Variacnce of measurments:
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:23]: Variance

File: test/navigation/kalman_filter.cpp [hunk 6]
```diff
@@ -10,21 +11,113 @@
 namespace hyped::test {
 
 TEST(KalmanFilter, construction)
+
+// Test simulating simple cart movement
+
+// Acceleration: 20ms^-2
+// Initial velocity: 0
+// Initial position: 0
+// Measurments taken every 1s
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:20]: Measurements

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
looks good, a couple of nitpicks

**Review by DylanCavers — APPROVED** [review]
lgtm

File: lib/navigation/filtering/kalman_filter.cpp [hunk 7]
```diff
@@ -22,18 +22,39 @@ KalmanFilter::KalmanFilter(const StateVector &initial_state,
...
+  // P_k = F*P_k-1*F^T + Q
+
+  error_covariance_
+    = transition_matrix * error_covariance_ * transition_matrix.transpose() + transition_covariance;
+
+  // Update
+
+  // K_k = P_k*H^T*(H*P_k*H^T + R)^-1
+
+  auto kalman_gain = error_covariance_ * measurement_matrix.transpose()
```
Comment (TomLonergan03) [inline lib/navigation/filtering/kalman_filter.cpp:41]: `const`

File: lib/navigation/filtering/kalman_filter.cpp [hunk 8]
```diff
@@ -22,18 +22,39 @@ KalmanFilter::KalmanFilter(const StateVector &initial_state,
...
+
+  // x_k = x_k + K_k*(z_k - H*x_k)
+
   state_estimate_
-    = prop_state_estimate + kalman_gain * (measurement - measurement_matrix * prop_state_estimate);
+    = state_estimate_ + kalman_gain * (measurement - measurement_matrix * state_estimate_);
+
+  // P_k = (I - K_k*H)*P_k*(I - K_k*H)^T + K_k*R*K_k^T
+
+  auto difference = Eigen::Matrix<core::Float, state_dimension, state_dimension>::Identity()
```
Comment (TomLonergan03) [inline lib/navigation/filtering/kalman_filter.cpp:53]: `const`

File: test/navigation/kalman_filter.cpp
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:1]: pretty much every variable in here should be `const` unless it has to be mutable

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/navigation/filtering/kalman_filter.cpp [hunk 9]
```diff
@@ -22,18 +22,39 @@ KalmanFilter::KalmanFilter(const StateVector &initial_state,
 }
 
 void KalmanFilter::filter(const MeasurementVector &measurement, const ControlInput &control_input)
+
```
Comment (TomLonergan03) [inline lib/navigation/filtering/kalman_filter.cpp:25]: remove this whitespace

File: test/navigation/kalman_filter.cpp [hunk 10]
```diff
@@ -9,22 +9,114 @@
...
+// Variance of measurments:
+//  Distance: 10
+//  Velocity: 5
+//  Acceleration: 3
+
+// Expected displacement: 4000m
+// Expected velocity: 400ms^-1
+
+TEST(KalmanFilter, cartRunSimulation)
+
```
Comment (TomLonergan03) [inline test/navigation/kalman_filter.cpp:29]: and this

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
Real nitpicks but they will also rerun the linter which didn't run on the last push for some reason

**Review by TomLonergan03 — APPROVED** [review]
looks good

**Pull request merged**